src/hazardous/base.rs

```diff
@@ -0,0 +1,156 @@
+use crate::errors::UnknownCryptoError;
+use std::{convert::TryFrom, fmt, marker::PhantomData};
+
+/// Marker trait for when a type contains some sensitive information.
+pub trait Secret {}
+
+/// Marker trait for when a type contains only non-sensitive information.
+/// Be careful if implementing this trait on your own. It cannot
+/// cause memory unsafety, and so is not marked `unsafe`. Implementing
+/// it can, however, lead to data types containing sensitive data ending
+/// up with APIs meant only for types containing only non-sensitive data.
+pub trait Public {}
+
+/// A small trait containing static information about the minimum and
+/// maximum size (in bytes) of a type containing data.
+pub trait Bounded {
+    /// The largest number of bytes this type should be allowed to hold.
+    const MIN: Option<usize> = None;
+
+    /// The smallest number of bytes this type should be allowed to hold.
+    const MAX: Option<usize> = None;
+}
+
+/// A generic holder for types that are basically just a bag of bytes
+/// with extra semantic meaning and restriction on top. We parameterize
+/// over the byte storage with parameter `B`. We parameterize over the
+/// API-level semantics of the type with phantom type `K`.
+#[derive(Clone)]
+pub struct Data<B, K> {
+    bytes: B,
+    phantom: PhantomData<K>,
```

Comment:
This was probably overambitious. 

It ended up working out fine implementation-wise, but the generated documentation is pretty difficult to parse. By conditionally implementing "secret" methods like `unprotected_as_bytes` based on type parameter bounds, the "secret" methods are mixed in with the "public" methods (like `as_ref`) all in the same docs page. That's not great when compared to the [status quo](https://docs.rs/orion/0.17.1/orion/hazardous/stream/chacha20/struct.SecretKey.html).

As a solution, I think it would be best to split up `Data<B, K>` into `Secret<B, K>` and `Public<B,K>`. It's still necessary to have both of the type parameters: the first is for parameterizing over storage type (`Array` vs `Vec`) and the second is for labeling the data with its purpose, like `Secret<Array<32>, Argon2iKey>`, which precents key misuse by making it a distinct type.

Reply:
I agree that splitting `Data<B, K>` into `Secret<B, K>` and `Public<B,K>` is probably the best choice here.